deps/jemalloc/src/jemalloc.c

```diff
@@ -3927,5 +3927,133 @@ jemalloc_postfork_child(void) {
+    szind_t szind;
+    bool is_slab;
+    rtree_szind_slab_read(tsdn, &extents_rtree, rtree_ctx, (uintptr_t)ptr, true, &szind, &is_slab);
+    if (likely(is_slab)) {
+        /* Small allocation. */
+        extent_t *slab = iealloc(tsdn, ptr);
+        arena_t *arena = extent_arena_get(slab);
+        szind_t binind = extent_szind_get(slab);
+        unsigned binshard = extent_binshard_get(slab);
+        bin_t *bin = &arena->bins[binind].bin_shards[binshard];
+        malloc_mutex_lock(tsdn, &bin->lock);
+        /* Don't bother moving allocations from the slab currently used for new allocations */
+        if (slab != bin->slabcur) {
+            if (bin->highest_slab_to_retain_inited && extent_snad_comp(slab, &bin->highest_slab_to_retain) > 0) {
+                defrag = 1;
+            }
+        }
+        malloc_mutex_unlock(tsdn, &bin->lock);
+    }
+    return defrag;
+}
+
+static long long ustime(void) {
+    struct timespec ts;
+    int res = clock_gettime(CLOCK_MONOTONIC, &ts);
+    assert(res == 0);
+    return (ts.tv_sec * 1000000) + (ts.tv_nsec / 1000);
+}
+
+static void init_defrag_bin_step(bin_t *bin, long long start_time, long long max_time) {
+    for (;bin->defrag_slabs_to_retain > 0; bin->defrag_slabs_to_retain--) {
+        extent_t *slab = extent_heap_remove_first(&bin->slabs_nonfull);
+        if (slab == NULL) {
+            bin->defrag_slabs_to_retain = 0;
+            break;
+        }
+
+        extent_heap_insert(&bin->slabs_nonfull_temp, slab);
+        if (!bin->highest_slab_to_retain_inited || extent_snad_comp(slab, &bin->highest_slab_to_retain) > 0) {
+            bin->highest_slab_to_retain = *slab;
+            bin->highest_slab_to_retain_inited = true;
+        }
+        if (ustime() - start_time >= max_time)
+            break;
+    }
+
+    if (bin->defrag_slabs_to_retain == 0) {
+        extent_heap_merge(&bin->slabs_nonfull, &bin->slabs_nonfull_temp);
+        bin->initing_defrag = false;
+    }
 }
+
+JEMALLOC_EXPORT void JEMALLOC_NOTHROW
+init_defrag(void) {
+    unsigned long long i;
+    tsd_t *tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
+
+    for (i = 0; i < SC_NBINS; i++) {
+        unsigned binshard;
+        bin_t *bin = arena_bin_choose_lock(tsd_tsdn(tsd), arena, i, &binshard);
+        const bin_info_t *bin_info = &bin_infos[i];
+        uint64_t used_slabs = bin->stats.nonfull_slabs + (bin->slabcur ? 1 : 0);
+        uint64_t full_slabs = bin->stats.curslabs - bin->stats.nonfull_slabs - (bin->slabcur ? 1 : 0);
+        uint64_t needed_slabs = (bin->stats.curregs - full_slabs * bin_info->nregs) / bin_info->nregs;
+        if ((bin->stats.curregs - full_slabs * bin_info->nregs) % bin_info->nregs) needed_slabs += 1;
+        needed_slabs += full_slabs;
+        used_slabs += full_slabs;
+        assert(bin->highest_slab_to_retain_inited == false);
+        assert(extent_heap_empty(&bin->slabs_nonfull_temp));
+        if (used_slabs != needed_slabs) {
```

Comment:
If there aren't many more used slabs than needed slabs then the init phase should be quick. So I'm not sure there's much value in this additional threshold.

Reply:
You are right, but the original and also this algorithm mitigates this by requiring the user to set relevant/sane `active-defrag-threshold-lower` and `active-defrag-ignore-bytes`. I'm not sure we need another threshold here, specifically this will be a rather difficult one to explain.